### DISCUSSION TITLE
Heap or Stack Allocated?

### DISCUSSION THREAD

**Discussion by pixmusix** [nkaaf/Arduino-List#35] [Q&A] [answered]
Hello.

Firstly, excellent library! Thanks :orange_heart: .

I'm wondering if, when using list, I need to de-allocate or destroy the list or it's contents?
Is everything stack allocated under the hood?

Thanks.

**Comment by nkaaf** [comment, marked as answer]
Hi, thanks for the feedback :smile: 

### Memory Allocation
This library is using dynamic memory allocation with ```malloc``` and ```new```, therefore the raw data of the lists are saved on the heap (internal references to the content on the stack). The list is saved on the stack or the heap, depending on your creation of the list (```new``` -> heap; static creation (like examples) -> stack).

#### ```new```/```delete```
The approach of using ```new``` is working in the context of an Arduino program (because of a custom override of ```new``` and ```delete```) or standard C++ program (and probably more software architectures which I haven't testes). It is not working when avr-gcc is used (e.g. in Microchip Studio). Until I dealt with your question, I was not aware of the fact with avr-gcc. Maybe I will look at this, if I have more time or someone needs it ;)

### Memory Freeing
There are some functions which return pointers, that you have to ```free``` after you aren't using the pointer anymore (look at the documentation of the functions you are using). I have free'd and deleted internally everything I was aware of (doing some tests with Valgrind). So you don't have to delete/free the content of the list manually. It is implemented in the clear/remove functions of the library. But if you are using a mutable list (which is not the default behaviour and must be "activated" by a constructor parameter), you are responsible for the content, because only pointer to your own variables are saved in the list. The freeing of the list object depends on the allocation as I mentioned in the first part of this comment (heap -> ```delete```; stack -> automatically deleted on scope exit).

### Discussion "Dynamic allocation in embedded systems"
To be honest I don't what to enter into the discussion (like in other forums). I made this library to offer a user-friendly implementation of lists which can be easily used on microcontrollers *and* normal business applications. It is maybe not the size efficient solution, but I was planning a "MinimalList" to implement a size efficient list to the library too (in the future). But keep in mind, that dynamic allocation can overflow your available memory, especially if you are using many items in lists, maybe multiple lists and more libraries and own code.

I hope I have answered your question and summarized the topic correctly.
Thank you very much for the question, because it made me dig deeper into the topic. Feel free if you have more questions :smile:

**Comment by pixmusix** [comment]
Thanks @nkaaf for taking the time to reply.

 > I have free'd and deleted internally everything I was aware of.

Amazing. Your work has paid off. Working through the examples I have found Arduino List to be super comfortable to use.

 > It is implemented in the clear/remove functions of the library.

On [this example](https://github.com/nkaaf/Arduino-List/blob/main/examples/List/ManageElements/ManageElements.ino) you have used this clear() func. That makes sense to me in the context of the setup() {} function, which is run once.

In the case of the loop() {}, when shall I use the clear() func? If the Arduino receives an unexpected power outage or I fail to catch a powerdown and do not implement clear(), will I have cause a leak?

**Reply by nkaaf** [reply]
Hi, sorry for the late message.

> Amazing. Your work has paid off. Working through the examples I have found Arduino List to be super comfortable to use.

Then I have done everything right :smile: 

The ```clear``` method is to delete all entries of the list. The ```remove``` method deletes only one entry at a given index. The examples are just to show the possibilities and functions of the library.

If your program ends, or you use a list only locally in a function, then you don't necessarily have to make sure that everything is empty at the end. The library takes care of that (internally it simply calls ```clear```).
If the Arduino gets an unexpected power dip and goes off, then all data is cleared from RAM. All data/variables from the programs are stored in the RAM. To store data persistently the EEPROM or FLASH can be used. These data remain even after a long time without power.

To answer your questions about the leak: You will not have a data leak in the classical sense of IT security. The data will be erased after a power dip, regardless of whether it was previously cleared or not (applies to RAM). However, if you store your data in EEPROM, then you have to consider for yourself if you want to clear the data by some mechanism.

**Reply by pixmusix** [reply]
> The data will be erased after a power dip, regardless of whether it was previously cleared or not (applies to RAM). However, if you store your data in EEPROM, then you have to consider for yourself if you want to clear the data by some mechanism.

Beautiful. This makes intuitive sense to me. I guess I just need to get a bit more comfortable with embedded systems.
Thank you for sharing your software. It's a wonderful thing. 😄
